### PR TITLE
Add documentation for incompatible change process.

### PR DESCRIPTION
**Pull request by dslomov**
Change-Id: I6129e587ffde66fd7d4ba26c147ae7c12409f3d2

### CODE REVIEW

File: site/docs/backward-compatibility.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+incompatible and require some changes from Bazel users.
+
+## At a glance
+
+1. Every breaking change is guarded with an `--incompatible_*` flag.
+1. Newly introduced incompatible flags default to off.
+1. For every `--incompatible_*` flag we have a github issue explaining
+   the change in behavior and giving a migration recipe.
+1. The migration window is at least one release long and is set by the author of the incompatible change.
+1. We announce what set of flags we flip with the next release one release in advance.
```
Comment (laurentlb) [inline site/docs/backward-compatibility.md]: I think the promise should be: We announce the flags we've switched in the release announcement.

It's hard to know for sure what will be switched, and we sometimes have to postpone changes. Users can know the plans by looking at GitHub labels.

File: site/docs/backward-compatibility.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+1. APIs and behavior guarded by an `--experimental_*` flag can change at any time.
+1. Users should never run their production builds with `--experimental_*`  or `--incompatible_*` flags.
```
Comment (laurentlb) [inline site/docs/backward-compatibility.md]: I don't think we should be so strict. Users can do what they want at their own risk (which is the item above).

File: site/docs/backward-compatibility.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+* Starlark language and APIs
+* Rules bundled with Bazel
+* Bazel APIs such as Remote Execution APIs or Build Event Protocol
+* Flags and their semantics
+
+## Incompatible changes and migration recipes
+
+When we introduce an incompatible change, we try to make it easier for Bazel
+users to updatre their code. We do this by means of _migration windows_ and
```
Comment (laurentlb) [inline site/docs/backward-compatibility.md]: "update"

File: site/docs/backward-compatibility.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+_migration recipes_.
+
+Migration window is one or more release of Bazel during which a migration from
+old funcitonality to new functionality is possible, according to a migration
```
Comment (laurentlb) [inline site/docs/backward-compatibility.md]: functionality

File: site/docs/backward-compatibility.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+recipe.
+
+During the migration window, both the old funcitonality and the new functionality
```
Comment (laurentlb) [inline site/docs/backward-compatibility.md]: functionality

File: site/docs/skylark/backward-compatibility.md [hunk 6]
```diff
@@ -1,15 +1,12 @@
 ---
 layout: documentation
-title: Backward Compatibility
+title: Starlark Incompatible CHnages
```
Comment (laurentlb) [inline site/docs/skylark/backward-compatibility.md]: This page also contains rule specific changes.

We should decide what we want to do with this page: either mention all incompatible changes, or kill it. The current state is misleading.

Comment (dslomov) [reply site/docs/skylark/backward-compatibility.md]: What would you like to do?

File: site/docs/backward-compatibility.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+# Backward Compatibility
+
+Bazel is evolving, and we will make changes to Bazel that at times will be
+incompatible and require some changes from Bazel users.
+
+## At a glance
+
+1. Every breaking change is guarded with an `--incompatible_*` flag.
+1. Newly introduced incompatible flags default to off.
+1. For every `--incompatible_*` flag we have a github issue explaining
```
Comment (jin) [inline site/docs/backward-compatibility.md]:
```suggestion
1. For every `--incompatible_*` flag we have a GitHub issue explaining
```

File: site/docs/backward-compatibility.md (hunk 2 above)
Comment (jin) [reply site/docs/backward-compatibility.md]: Maybe "We do not recommend running their production builds"?

File: site/docs/backward-compatibility.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+When we introduce an incompatible change, we try to make it easier for Bazel
+users to update their code. We do this by means of _migration windows_ and
+_migration recipes_.
+
+Migration window is one or more release of Bazel during which a migration from
+old functionality to new functionality is possible, according to a migration
+recipe.
+
+During the migration window, both the old functionality and the new functionality
```
Comment (jin) [inline site/docs/backward-compatibility.md]: Could we add something for the case where a migration recipe is not available? For example, the [aapt2 migration](https://github.com/bazelbuild/bazel/issues/6907) is not doable mechanically most of the time, but we keep another flag around to switch to the old behavior.

File: site/docs/backward-compatibility.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+1. After the code is migrated, the users can check whether they migrated
+   successfully by building with `--incompatible_foo=true`. However, their
+   code will continue to work in the same release in default state (where
+   `--incompatible_foo` is of), as well after the migration window is over
+   (at which point the flag will be effectively on).
+
+## Communicating incompatible changes
+
+The primary source of information about incompatible changes are GitHub issues
+marked with an ["incompatible-change" label](https://github.com/bazelbuild/bazel/issues?q=label%3Aincompatible-change+).
```
Comment (jin) [inline site/docs/backward-compatibility.md]:
```suggestion
marked with an ["incompatible-change" label](https://github.com/bazelbuild/bazel/issues?q=label%3Aincompatible-change).
```

File: site/docs/backward-compatibility.md (hunk 2 above)
Comment (dslomov) [reply site/docs/backward-compatibility.md]: We have to make it very clear that especially the owners of things that are dependent upon widely (such as rulesets) _should not_ expect any experimental or incompatible flags to be present.

I'd rather keep this strong language here.

File: site/docs/backward-compatibility.md (hunk 1 above)
Comment (dslomov) [reply site/docs/backward-compatibility.md]: Changed the  language to "intent to flip". The point of this item is that the users can know ahead of time what's coming

**Comment by dslomov** [comment]
Ready for another look

**Comment by dslomov** [comment]
PTAL

File: site/docs/skylark/backward-compatibility.md [hunk 10]
```diff
@@ -1,34 +1,16 @@
 ---
 layout: documentation
-title: Backward Compatibility
+title: Starlark Incompatible Chnages
```
Comment (laurentlb) [inline site/docs/skylark/backward-compatibility.md]:
```suggestion
title: Incompatible Changes
```

(remove Starlark + fix typo)

**Comment by dslomov** [comment]
@jin and @spomorski looks ok to you too?

**Pull request merged**